### PR TITLE
draft blog post for writing a toy optimizer

### PR DESCRIPTION
**Pull request by cfbolz**

### CODE REVIEW

File: posts/2022/07/toy-optimizer.rst [hunk 1]
```diff
@@ -0,0 +1,912 @@
...
+.. category:
+.. link:
+.. description:
+.. type: rest
+.. author: Carl Friedrich Bolz-Tereick
+
+
+In this blog post I want to show the complete code (in Python3) of how a very
+simple optimizer for sequences of operations can work. These algorithms could
+be part of a (really simple) compiler, or a JIT.
```
Comment (mattip) [inline posts/2022/07/toy-optimizer.rst:14]: Could you add another sentence about motivation: is this a tutorial on what the PyPy JIT does and at what stage, is this a general JIT theory tutorial, is this the first in a series of "How JITs works", ?

File: posts/2022/07/toy-optimizer.rst [hunk 2]
```diff
@@ -0,0 +1,912 @@
...
+snippets of IR by hand. I also won't talk about what happens after the
+optimization (usually the optimized IR is translated into machine code).
+
+
+Implementing the Intermediate Representation
+=============================================
+
+Let's start modelling the intermediate representation with Python classes.
+First we define a base class of all values that can be used as arguments in
+operations, and let's also add a class that represents constants:
```
Comment (mattip) [inline posts/2022/07/toy-optimizer.rst:60]:
```suggestion
operations, and let's also add a dataclass_ that represents constants:
.. _dataclass: https://realpython.com/python-data-classes/
```

File: posts/2022/07/toy-optimizer.rst [hunk 3]
```diff
@@ -0,0 +1,912 @@
...
+    @dataclass(eq=False)
+    class Constant(Value):
+        value : Any
+
+
+One consequence of the fact that every variable is assigned to only once is
+that variables are in a one-to-one correspondence with the right-hand-side of
+their unique assignments. That means that we don't need a class that represents
+variables at all. Instead, it's sufficient to have a class that represents an
+operation, and that is the same as the variable that it defines:
```
Comment (mattip) [inline posts/2022/07/toy-optimizer.rst:83]:
```suggestion
operation (the right-hand side), and that by definition is the same as the variable (left-hand side) that it defines:
```

File: posts/2022/07/toy-optimizer.rst [hunk 4]
```diff
@@ -0,0 +1,912 @@
...
+
+It's impossible to see what is going on here, because the `Operations` in the
+basic block appear several times, once as elements of the list but then also as
+arguments to operations further down in the list. So we need some code that
+turns things back into a readable textual representation, so we have a chance
+to debug.
+
+.. code:: python
+
+    def bb_to_str(l : Block, varprefix : str = "var"):
```
Comment (mattip) [inline posts/2022/07/toy-optimizer.rst:238]: `l` is not the best choice for an argument name, it took me quite a while to figure out where it was used when scanning the code
```suggestion
    def bb_to_str(l : Block, varprefix : str = "var"):
```

File: posts/2022/07/toy-optimizer.rst [hunk 5]
```diff
@@ -0,0 +1,912 @@
...
+
+    optvar0 = getarg(0)
+    optvar1 = getarg(1)
+    optvar2 = add(optvar1, 17)
+    optvar3 = mul(optvar0, optvar2)
+    optvar4 = add(optvar3, optvar2)
+
+We left out the second ``add`` (which defines ``v4``), and then replaced the
+usage of ``v4`` with ``v2`` in the final operation.
+
```
Comment (mattip) [inline posts/2022/07/toy-optimizer.rst:340]: maybe add line numbers, and then you can say "We left out line ... and replaced v4 with v2 in line ...

File: posts/2022/07/toy-optimizer.rst [hunk 6]
```diff
@@ -0,0 +1,912 @@
...
+.. _`most important optimization`: https://www.pypy.org/posts/2010/09/escape-analysis-in-pypys-jit-1780048403046080197.html
+
+
+Some Further Pointers
+======================
+
+This is a blog post, not a paper, but I still want to give some pointers to
+literature that can be read to understand the concepts that are introduced here
+in much bigger generality.
```
Comment (mattip) [inline posts/2022/07/toy-optimizer.rst:872]: No need to demean this blog post, summarizing is sometimes harder than writing a paper.
```suggestion
This is a short introduction to many fundamental concepts that are
covered in more general literature.
```

File: posts/2022/07/toy-optimizer.rst (hunk 1 above)
Comment (mattip) [reply posts/2022/07/toy-optimizer.rst:14]: This is touched on near the end
> And indeed, what I describe in this post is very similar to the one
PyPy's JIT optimizer uses

so maybe rephrase it here:
> and in fact the optimizer style we end up with is very similar to the one in PyPy's JIT

**Comment by mattip** [comment]
I made a pass over the blog. Looks great, I like the way the simple concepts are used to build a complete story.

File: posts/2022/07/toy-optimizer.rst (hunk 5 above)
Comment (cfbolz) [reply posts/2022/07/toy-optimizer.rst:345]: yeah, that would be great, unfortunately the style of numbered blocks is really different than regular ones :-(

image

so I'm inclined to leave it as it is.

File: posts/2022/07/toy-optimizer.rst (hunk 2 above)
Comment (cfbolz) [reply posts/2022/07/toy-optimizer.rst:65]: removed the data classes entirely

File: posts/2022/07/toy-optimizer.rst (hunk 1 above)
Comment (cfbolz) [reply posts/2022/07/toy-optimizer.rst]: thanks, did this

File: posts/2022/07/toy-optimizer.rst (hunk 4 above)
Comment (cfbolz) [reply posts/2022/07/toy-optimizer.rst]: I renamed the argument to bb like everywhere else

File: posts/2022/07/toy-optimizer.rst (hunk 6 above)
Comment (cfbolz) [reply posts/2022/07/toy-optimizer.rst]: makes sense, wrote something along these lines
